Address comment related language support (#92)
ietf-wg-dnsop/draft-ietf-dnsop-structured-dns-error: draft-ietf-dnsop-structured-dns-error.md

l: (language)
: The "l" field indicates the language used for the JSON-encoded "j" and "o" fields.  The value of this field MUST conform to the
  language tag syntax specified in {{Section 2.1 of !RFC5646}}.

> : This field is optional but MUST be conveyed when the "j" or "o" fields, or both, are present.

Reviewer: ```suggestion
: This field is optional but MUST be included when either the "j" or "o" fields are present.
```

: The integer values are defined in the IANA-managed registry for DNS Sub-Error Codes in {{IANA-SubError}}.
: This field is optional.

When multiple blocking causes apply simultaneously
(e.g., a domain is blocked for both malware and phishing reasons),
a single SDE response is returned. The "s" field MUST convey the
primary blocking cause. The "j" field MUST be used to provide
additional context describing all applicable causes.

o: (organization)
: 'UTF-8'-encoded human-friendly name of the organization that filtered this particular DNS query.
: This field is optional.

l: (language)
: The "l" field indicates the language used for the JSON-encoded "j" and "o" fields.  The value of this field MUST conform to the
  language tag syntax specified in {{Section 2.1 of !RFC5646}}.

> : This field is optional but MUST be included when either the "j" or "o" fields are present.

Reviewer 2: Then it's not optional.

Author: I will fix the text as follows:
This field is optional. It MUST be included when either the "j" or
"o" fields are present.

field of the EDE option is encoded and processed in accordance with
this specification.

A client that wishes to express a preferred response language
MUST populate the OPTION-DATA of the SDE option with an ordered
list of RFC 5646 {{!RFC5646}} language tags, listed from most to
least preferred, using the format defined in {{SDE}}. The list

> SHOULD contain no more than 4 entries and MUST NOT contain more than

Reviewer: Please add justification / explanation for the 4- and 8-entry limit, so implementors know why there is a limit.  (Probably just needs to point at server section to explain why there is a limit.)

Author: I think users typically configure 2 languages in their browser. Since clients may expand each language tag into a language-region pair (e.g., "en-US" → "en-US, en"), the list doubles to at most 4 entries, hence the SHOULD ≤4 limit. The MUST ≤8 accommodates the rare case of a user with up to 4 configured languages after expansion.

You may want to look into https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Accept-Language.

Reviewer: suggest:

  "to accommodate two language with two language tags per language (e.g., American english often uses the two tags "en-US" and "en"), the list SHOULD contain no more than 4 entries ..."

8 entries. A client that has no language preference MUST set

> OPTION-LENGTH to 0.

Reviewer: The MUST constrains future use of OPTION-DATA.

And this is merely a 'hint'; maybe want to add that it is just a hint?

how about:

```suggestion
8 entries. A client that has no language preference sends no OPTION-DATA.
```

But overall, I am not comfortable adding language:  it further complicates DNS caching which now has to be language-aware, too.

Author: Good point, I will add it as a limitation.

field of the EDE option is encoded and processed in accordance with
this specification.

A client that wishes to express a preferred response language

> MUST populate the OPTION-DATA of the SDE option with an ordered

Reviewer 2: Which client would want to do that?

Author: Browsers that do not wish to express a language preference can set OPTION-LENGTH to 0, which is the default. Non-browser clients can include their preferred language list to receive a localized response. For example, in Enterprise DNS filtering deployments, structured error data is logged and presented to administrators who may prefer a specific language.

list of RFC 5646 {{!RFC5646}} language tags, listed from most to
least preferred, using the format defined in {{SDE}}. The list
SHOULD contain no more than 4 entries and MUST NOT contain more than
8 entries. The limit of 4 entries reflects
typical real-world language preference lists, as a user with 2
configured languages may generate up to 4 entries after
language-only fallback tags are added (e.g., "en-US, en, zh-CN,
zh"). The hard limit of 8 entries bounds the contribution of
the SDE option to the DNS query size (see {{?RFC9715}}) and ensures
predictable server processing as described in {{server-response}}.

> A client that has no language preference MUST set OPTION-LENGTH to 0.

Reviewer 3: ```suggestion
A client that has no language preference MUST set OPTION-LENGTH to 0.
This is the default behavior.
```

For privacy reasons, clients MAY send fewer language

> entries than the user has configured, or omit the language list

Reviewer 3: ```suggestion
entries than the user has configured or omit the language list
```

> entirely by setting OPTION-LENGTH to 0.

Reviewer 2: How is this helping privacy? I'd still expose a valid subset. If anything, I'd request ALL languages so the server can't tell which I want (which clearly won't work).

Author: You may want to look into https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Accept-Language, The specific text is:

For privacy purposes (reducing fingerprinting), both Accept-Language and navigator.languages may not include the full list of user preferences. For example, in Safari (always) and Chrome's incognito mode, only one language is listed.

For clarity, will fix text as follows:

For privacy reasons, clients MAY send only one language entry or
omit the language list entirely by setting OPTION-LENGTH to 0.

The "l" field in the JSON body of the EXTRA-TEXT response
({{name-spec}}) remains the authoritative indicator of the language

> actually used in the "j" and "o" fields and MUST continue to be

Reviewer 3: ```suggestion
used in the "j" and "o" fields and MUST continue to be
```

> used by clients for localisation and machine translation decisions.

Reviewer 2: Clients will not be bound by this, whatever you write here.

## Server Generating Response {#server-response}

When the DNS server filters its DNS response to a
query (e.g., A or AAAA resource record query), the DNS response MAY contain an empty answer, NXDOMAIN, or (less
ideally) forged response, as desired by the DNS
server.

If the query contained the SDE EDNS option ({{client-request}}), and the DNS server returns an EDE code of "Blocked", "Filtered", "Censored", or "Blocked by Upstream DNS Server", the DNS server SHOULD include additional detail in the EXTRA-TEXT field encoded as structured and machine-readable data in accordance with the present specification, unless configured otherwise. If including the additional detail would cause the response to exceed the EDNS0 size {{?RFC9715}} (and thus setting TC=1), the server SHOULD first attempt to reduce the response size by omitting the "j" and "o" fields before omitting the EXTRA-TEXT entirely. In deployments using DoT, DoH, or DoQ, transport size limitations are unlikely to necessitate omission of structured data in the EXTRA-TEXT field.

> If the SDE option OPTION-DATA is non-empty, and the server intends

Reviewer 3: ```suggestion
If the SDE option OPTION-DATA is non-empty and the server intends
```

If the query contained the SDE EDNS option ({{client-request}}), and the DNS server returns an EDE code of "Blocked", "Filtered", "Censored", or "Blocked by Upstream DNS Server", the DNS server SHOULD include additional detail in the EXTRA-TEXT field encoded as structured and machine-readable data in accordance with the present specification, unless configured otherwise. If including the additional detail would cause the response to exceed the EDNS0 size {{?RFC9715}} (and thus setting TC=1), it SHOULD be omitted. In deployments using DoT, DoH, or DoQ, transport size limitations are unlikely to necessitate omission of structured data in the EXTRA-TEXT field.

If the SDE option OPTION-DATA is non-empty, and the server intends
to populate the "j" or "o" fields, the server MUST perform
{{!RFC4647}} lookup matching against the language entries in the
order they appear, selecting the first entry for which localised
text is available. If a match is found, the server SHOULD populate 

> the "j" and "o" fields in the matched language; a field MAY be 

Reviewer: "a field MAY be omitted" --> I don't know what is intended by "a field".  Does it mean a *language* may be omitted??

Author: Fixed text below: 
If the SDE option OPTION-DATA is non-empty, and the server intends to populate the "j" or "o" fields, the server MUST perform {{!RFC4647}} lookup matching against the language entries in the order they appear, selecting the first entry for which localised text is available. If a match is found, the server SHOULD populate the "j" and "o" fields in the matched language; either field MAY be omitted if the server has no value to convey for it. If either field is present, the server MUST set the "l" field to the matched language tag. If no match is found, the server MUST fall back to its default language. A failure to match a preferred language MUST NOT prevent the server from returning a response.

omitted if the server has no value to convey for it. If either field 
is present, the server MUST set the "l" field to the matched language 
tag. If no match is found, the server MUST fall back to its default 
language. A failure to match a preferred language MUST NOT prevent 
the server from returning a response.

A server receiving an SDE option with unrecognised or malformed
OPTION-DATA MUST silently ignore the OPTION-DATA and process the
option as if OPTION-LENGTH were 0.

> 

Reviewer: Does the spec need to explain strategy for server to shorten text (e.g., "malware present for 23 days" -> "malware 23d") to attempt to get more languages to fit, or that it should just truncate in the middle of the text (e.g., "malware present f") because the response payload was consumed by the request's most-preferred French or whatever.

Or maybe the strategy doesn't matter for interop?

Author: The server would only provide response in one language. The spec already requires JSON values to be as short as possible and allows the server to omit fields from EXTRA-TEXT if the response would exceed the EDNS0 size limit .

indicate support for I-JSON encoding in the EXTRA-TEXT field of an
Extended DNS Error (EDE) option.

The SDE option MAY carry an OPTION-DATA field containing an ordered
list of preferred languages. The OPTION-LENGTH field indicates the
total length of the OPTION-DATA in octets. An OPTION-LENGTH of 0
indicates no language preference and is semantically equivalent to
the absence of OPTION-DATA.

The OPTION-DATA, when present, has the following format:

~~~
+--+--+--+--+--+--+--+--+
|        LANG-COUNT      |   1 octet
+--+--+--+--+--+--+--+--+
|        LANG-LEN[0]     |   1 octet
+--+--+--+--+--+--+--+--+
|        LANG-TAG[0]     |   LANG-LEN[0] octets (RFC 5646)
+--+--+--+--+--+--+--+--+
|        ...             |   repeat LANG-LEN / LANG-TAG

> +--+--+--+--+--+--+--+--+

Reviewer: one byte more efficient would be separating LANG-TAG with an ASCII separator, such as comma or colon or suchlike.  That would allow eliminating the initial LANG-COUNT.  I also don't like the initial LANG-COUNT because what if it says "8", but there are only 2 languages listed and the rest of the OPTION-DATA is, say 0xFF (egads, we could run off end of the buffer) or 0x00.

Considering the response is JSON-encoded, should the language preference of the request perhaps also be JSON-encoded?  If not, why not.

Author: Good point, I will go with design 1 (use ASCII separator).

Reference:
: RFC XXXX

Note: The OPTION-DATA for this option carries an optional ordered

> list of preferred languages as defined in {{SDE}} of this document.

Reviewer 3: ```suggestion
list of preferred languages as defined in {{SDE}}.
```
